File: dotnet/src/Connectors/Connectors.AI.MultiConnector/PromptMatching/HybridDictionary.cs
// Copyright (c) MyIA. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;

namespace MyIA.SemanticKernel.Connectors.AI.MultiConnector.PromptMatching
{
    /// <summary>
    /// Dictionnaire hybride qui bascule entre un tableau et un dictionnaire en fonction du nombre d'éléments.
    /// Optimisé pour les petites collections avec un accès rapide.
    /// </summary>
    /// <typeparam name="K">Type de clé</typeparam>
    /// <typeparam name="V">Type de valeur</typeparam>
    public class HybridDictionary<K, V> where K : notnull
    {
        // Seuil à partir duquel on bascule d'un tableau à un dictionnaire
        private const int DefaultThreshold = 7;

        private readonly int _threshold;
        private readonly IEqualityComparer<K> _comparer;

        // Stockage sous forme de liste de paires clé-valeur pour les petites collections
        private List<KeyValuePair<K, V>>? _list;

        // Stockage sous forme de dictionnaire pour les grandes collections
        private Dictionary<K, V>? _dictionary;

        /// <summary>
        /// Nombre d'éléments dans la collection
        /// </summary>
        public int Count => _list != null ? _list.Count : _dictionary?.Count ?? 0;

        /// <summary>
        /// Constructeur par défaut
        /// </summary>
        public HybridDictionary() : this(DefaultThreshold, EqualityComparer<K>.Default)
        {
        }

        /// <summary>
        /// Constructeur avec seuil personnalisé
        /// </summary>
        /// <param name="threshold">Seuil à partir duquel on bascule d'un tableau à un dictionnaire</param>
        public HybridDictionary(int threshold) : this(threshold, EqualityComparer<K>.Default)
        {
        }

        /// <summary>
        /// Constructeur avec comparateur personnalisé
        /// </summary>
        /// <param name="comparer">Comparateur d'égalité pour les clés</param>
        public HybridDictionary(IEqualityComparer<K> comparer) : this(DefaultThreshold, comparer)
        {
        }

        /// <summary>
        /// Constructeur avec seuil et comparateur personnalisés
        /// </summary>
        /// <param name="threshold">Seuil à partir duquel on bascule d'un tableau à un dictionnaire</param>
        /// <param name="comparer">Comparateur d'égalité pour les clés</param>
        public HybridDictionary(int threshold, IEqualityComparer<K> comparer)
        {
            _threshold = threshold > 0 ? threshold : DefaultThreshold;
            _comparer = comparer ?? EqualityComparer<K>.Default;
            _list = new List<KeyValuePair<K, V>>();
        }

        /// <summary>
        /// Indexeur pour accéder aux valeurs par clé
        /// </summary>
        /// <param name="key">Clé à rechercher</param>
        /// <returns>Valeur associée à la clé</returns>
        public V this[K key]
        {
            get
            {
                if (_dictionary != null)
                {
                    return _dictionary[key];
                }

                if (_list != null)
                {
                    foreach (var pair in _list)
                    {
                        if (_comparer.Equals(pair.Key, key))
                        {
                            return pair.Value;
                        }
                    }
                }

                throw new KeyNotFoundException($"La clé '{key}' n'a pas été trouvée dans le dictionnaire.");
            }
            set
            {
                if (_dictionary != null)
                {
                    _dictionary[key] = value;
                    return;
                }

                if (_list != null)
                {
                    for (int i = 0; i < _list.Count; i++)
                    {
                        if (_comparer.Equals(_list[i].Key, key))
                        {
                            _list[i] = new KeyValuePair<K, V>(key, value);
                            return;
                        }
                    }

                    _list.Add(new KeyValuePair<K, V>(key, value));

                    // Si le nombre d'éléments dépasse le seuil, on bascule vers un dictionnaire
                    if (_list.Count > _threshold)
                    {
                        ConvertToDict();
                    }
                }
            }
        }

        /// <summary>
        /// Vérifie si la clé existe dans le dictionnaire
        /// </summary>
        /// <param name="key">Clé à rechercher</param>
        /// <returns>True si la clé existe, false sinon</returns>
        public bool ContainsKey(K key)
        {
            if (_dictionary != null)
            {
                return _dictionary.ContainsKey(key);
            }

            if (_list != null)
            {
                foreach (var pair in _list)
                {
                    if (_comparer.Equals(pair.Key, key))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        /// <summary>
        /// Tente de récupérer la valeur associée à une clé
        /// </summary>
        /// <param name="key">Clé à rechercher</param>
        /// <param name="value">Valeur associée à la clé si trouvée</param>
        /// <returns>True si la clé existe, false sinon</returns>
        public bool TryGetValue(K key, out V value)
        {
            if (_dictionary != null)
            {
                return _dictionary.TryGetValue(key, out value!);
            }

            if (_list != null)
            {
                foreach (var pair in _list)
                {
                    if (_comparer.Equals(pair.Key, key))
                    {
                        value = pair.Value;
                        return true;
                    }
                }
            }

            value = default!;
            return false;
        }

        /// <summary>
        /// Ajoute une paire clé-valeur au dictionnaire
        /// </summary>
        /// <param name="key">Clé à ajouter</param>
        /// <param name="value">Valeur à associer à la clé</param>
        public void Add(K key, V value)
        {
            if (_dictionary != null)
            {
                _dictionary.Add(key, value);
                return;
            }

            if (_list != null)
            {
                foreach (var pair in _list)
                {
                    if (_comparer.Equals(pair.Key, key))
                    {
                        throw new ArgumentException($"Une entrée avec la même clé existe déjà: '{key}'");
                    }
                }

                _list.Add(new KeyValuePair<K, V>(key, value));

                // Si le nombre d'éléments dépasse le seuil, on bascule vers un dictionnaire
                if (_list.Count > _threshold)
                {
                    ConvertToDict();
                }
            }
        }

        /// <summary>
        /// Supprime une entrée du dictionnaire
        /// </summary>
        /// <param name="key">Clé à supprimer</param>
        /// <returns>True si la clé a été supprimée, false si elle n'existait pas</returns>
        public bool Remove(K key)
        {
            if (_dictionary != null)
            {
                return _dictionary.Remove(key);
            }

            if (_list != null)
            {
                for (int i = 0; i < _list.Count; i++)
                {
                    if (_comparer.Equals(_list[i].Key, key))
                    {
                        _list.RemoveAt(i);
                        return true;
                    }
                }
            }

            return false;
        }

        /// <summary>
        /// Efface toutes les entrées du dictionnaire
        /// </summary>
        public void Clear()
        {
            if (_dictionary != null)
            {
                _dictionary.Clear();
                _dictionary = null;
                _list = new List<KeyValuePair<K, V>>();
            }
            else if (_list != null)
            {
                _list.Clear();
            }
        }

        /// <summary>
        /// Retourne toutes les clés du dictionnaire
        /// </summary>
        public IEnumerable<K> Keys
        {
            get
            {
                if (_dictionary != null)
                {
                    return _dictionary.Keys;
                }

                if (_list != null)
                {
                    return _list.Select(pair => pair.Key);
                }

                return Enumerable.Empty<K>();
            }
        }

        /// <summary>
        /// Retourne toutes les valeurs du dictionnaire
        /// </summary>
        public IEnumerable<V> Values
        {
            get
            {
                if (_dictionary != null)
                {
                    return _dictionary.Values;
                }

                if (_list != null)
                {
                    return _list.Select(pair => pair.Value);
                }

                return Enumerable.Empty<V>();
            }
        }

        /// <summary>
        /// Convertit la liste en dictionnaire lorsque le seuil est dépassé
        /// </summary>
        private void ConvertToDict()
        {
            if (_list == null || _dictionary != null)
            {
                return;
            }

            _dictionary = new Dictionary<K, V>(_list.Count, _comparer);
            foreach (var pair in _list)
            {
                _dictionary.Add(pair.Key, pair.Value);
            }

            _list = null;
        }
    }
}

File: python/prompt_matching/core/hybrid_dictionary.py
"""
Module contenant l'implémentation du HybridDictionary.

Le HybridDictionary est une structure de données qui combine les avantages d'une liste et d'un dictionnaire,
avec un seuil à partir duquel elle change de comportement interne pour des raisons de performance.
"""

from typing import TypeVar, Generic, Dict, List, Tuple, Iterator, Optional, Callable, Any

K = TypeVar('K')  # Type de clé
V = TypeVar('V')  # Type de valeur

class HybridDictionary(Generic[K, V]):
    """
    Une structure de données qui combine les avantages d'une liste et d'un dictionnaire.
    
    Pour un petit nombre d'éléments, elle utilise une liste de tuples (clé, valeur) pour économiser
    de la mémoire. Une fois que le nombre d'éléments dépasse un seuil, elle bascule vers un
    dictionnaire standard pour de meilleures performances de recherche.
    """
    
    def __init__(self, threshold: int = 10, key_comparer: Callable[[K, K], bool] = None):
        """
        Initialise une nouvelle instance de HybridDictionary.
        
        Args:
            threshold: Seuil à partir duquel basculer vers un dictionnaire standard
            key_comparer: Fonction de comparaison des clés (si None, utilise l'égalité standard)
        """
        self._threshold = threshold
        self._key_comparer = key_comparer
        self._items_list: List[Tuple[K, V]] = []
        self._items_dict: Dict[K, V] = {}
        self._using_dict = False
    
    @property
    def count(self) -> int:
        """Nombre d'éléments dans le dictionnaire."""
        return len(self._items_dict) if self._using_dict else len(self._items_list)
    
    def add(self, key: K, value: V) -> None:
        """
        Ajoute une paire clé-valeur au dictionnaire.
        
        Args:
            key: La clé à ajouter
            value: La valeur associée à la clé
            
        Raises:
            ValueError: Si la clé existe déjà
        """
        if self.contains_key(key):
            raise ValueError(f"La clé '{key}' existe déjà dans le dictionnaire")
        
        if self._using_dict:
            self._items_dict[key] = value
        else:
            self._items_list.append((key, value))
            
            # Basculer vers un dictionnaire si le seuil est dépassé
            if len(self._items_list) > self._threshold:
                self._convert_to_dict()
    
    def __setitem__(self, key: K, value: V) -> None:
        """
        Définit ou remplace une valeur associée à une clé.
        
        Args:
            key: La clé à définir
            value: La valeur à associer à la clé
        """
        if self._using_dict:
            self._items_dict[key] = value
        else:
            for i, (k, _) in enumerate(self._items_list):
                if self._compare_keys(k, key):
                    self._items_list[i] = (key, value)
                    return
            
            # La clé n'existe pas, l'ajouter
            self._items_list.append((key, value))
            
            # Basculer vers un dictionnaire si le seuil est dépassé
            if len(self._items_list) > self._threshold:
                self._convert_to_dict()
    
    def __getitem__(self, key: K) -> V:
        """
        Récupère la valeur associée à une clé.
        
        Args:
            key: La clé à rechercher
            
        Returns:
            La valeur associée à la clé
            
        Raises:
            KeyError: Si la clé n'existe pas
        """
        if self._using_dict:
            return self._items_dict[key]
        else:
            for k, v in self._items_list:
                if self._compare_keys(k, key):
                    return v
            
            raise KeyError(f"La clé '{key}' n'existe pas dans le dictionnaire")
    
    def contains_key(self, key: K) -> bool:
        """
        Vérifie si une clé existe dans le dictionnaire.
        
        Args:
            key: La clé à rechercher
            
        Returns:
            True si la clé existe, False sinon
        """
        if self._using_dict:
            return key in self._items_dict
        else:
            return any(self._compare_keys(k, key) for k, _ in self._items_list)
    
    def try_get_value(self, key: K) -> Tuple[bool, Optional[V]]:
        """
        Tente de récupérer la valeur associée à une clé.
        
        Args:
            key: La clé à rechercher
            
        Returns:
            Un tuple (succès, valeur) où succès est True si la clé existe,
            et valeur est la valeur associée à la clé ou None si la clé n'existe pas
        """
        if self._using_dict:
            if key in self._items_dict:
                return True, self._items_dict[key]
            return False, None
        else:
            for k, v in self._items_list:
                if self._compare_keys(k, key):
                    return True, v
            return False, None
    
    def remove(self, key: K) -> bool:
        """
        Supprime une paire clé-valeur du dictionnaire.
        
        Args:
            key: La clé à supprimer
            
        Returns:
            True si la clé a été supprimée, False si elle n'existait pas
        """
        if self._using_dict:
            if key in self._items_dict:
                del self._items_dict[key]
                return True
            return False
        else:
            for i, (k, _) in enumerate(self._items_list):
                if self._compare_keys(k, key):
                    self._items_list.pop(i)
                    return True
            return False
    
    def clear(self) -> None:
        """Supprime tous les éléments du dictionnaire."""
        if self._using_dict:
            self._items_dict.clear()
        else:
            self._items_list.clear()
    
    @property
    def keys(self) -> List[K]:
        """Liste des clés du dictionnaire."""
        if self._using_dict:
            return list(self._items_dict.keys())
        else:
            return [k for k, _ in self._items_list]
    
    @property
    def values(self) -> List[V]:
        """Liste des valeurs du dictionnaire."""
        if self._using_dict:
            return list(self._items_dict.values())
        else:
            return [v for _, v in self._items_list]
    
    def _convert_to_dict(self) -> None:
        """Convertit la liste interne en dictionnaire."""
        self._items_dict = {k: v for k, v in self._items_list}
        self._items_list = []
        self._using_dict = True
    
    def _compare_keys(self, key1: K, key2: K) -> bool:
        """
        Compare deux clés en utilisant le comparateur personnalisé s'il existe.
        
        Args:
            key1: Première clé
            key2: Deuxième clé
            
        Returns:
            True si les clés sont égales, False sinon
        """
        if self._key_comparer is not None:
            return self._key_comparer(key1, key2)
        return key1 == key2
    
    def __iter__(self) -> Iterator[Tuple[K, V]]:
        """
        Renvoie un itérateur sur les paires clé-valeur du dictionnaire.
        
        Returns:
            Un itérateur sur les paires clé-valeur
        """
        if self._using_dict:
            return iter(self._items_dict.items())
        else:
            return iter(self._items_list)
    
    def __len__(self) -> int:
        """
        Renvoie le nombre d'éléments dans le dictionnaire.
        
        Returns:
            Le nombre d'éléments
        """
        return self.count